### ISSUE TITLE
Safari 12 issue: fields dynamically added at runtime are not positioned correctly

### ISSUE THREAD

**Issue by ZheSun88** [vaadin/vaadin-form-layout#110]
@martinisraelsen commented on [Tue Apr 09 2019](https://github.com/vaadin/flow/issues/5446)

When I add a FormLayout/FormItem dynamically, some of the last elements are not always positioned correctly. Resizing the screen will cause the fields to be repositioned to their correct location.

I've seen this happen in Safari on V12 and V13. I have not been able to reproduce the issue on other browsers. Vaadin V10 and V11 doesn't appear to have this bug either. It's possible that it's a Safari bug rather than a Vaadin bug, but then I would expect it to also happen consistently with V10 and V11.

I have attached a Vaadin starter project that has the error. To reproduce the error, run the project with jetty:run and open the app in Safari. Click "Add line" until a row appears without the select field and the button.

Also included a few screenshots etc.

![Screen Shot 2019-04-09 at 8 29 56 AM](https://user-images.githubusercontent.com/6271003/55809673-f18d7d80-5aab-11e9-8711-2575dfbc90ca.png)
![Screen Shot 2019-04-09 at 8 30 19 AM](https://user-images.githubusercontent.com/6271003/55809689-f6eac800-5aab-11e9-8c30-115b1f87fe2c.png)

![formlayout position error](https://user-images.githubusercontent.com/6271003/55809824-36191900-5aac-11e9-8baf-e85bfbe7cfd3.gif)
[Vaadin-Appstarter.zip](https://github.com/vaadin/flow/files/3059561/Vaadin-Appstarter.zip)

---

@martinisraelsen commented on [Sun Apr 14 2019](https://github.com/vaadin/flow/issues/5446#issuecomment-483041926)

Here's an updated app that exhibits both this issue and the layout issue reported in https://github.com/vaadin/flow/issues/4530 (which may or may not be the same).

To reproduce this layout issue, run with jetty, open in Safari. Click "Add User" followed by "Add User". The second time all fields are missing, but will show up if you resize the window or click inside one of the (missing) fields.

[Vaadin-Appstarter.zip](https://github.com/vaadin/flow/files/3077769/Vaadin-Appstarter.zip)
 

---

@martinisraelsen commented on [Wed May 08 2019](https://github.com/vaadin/flow/issues/5446#issuecomment-490612340)

@ZheSun88 Perhaps this issue should be logged in https://github.com/vaadin/vaadin-form-layout? Are you able to move it over?

---

@ZheSun88 commented on [Thu May 09 2019](https://github.com/vaadin/flow/issues/5446#issuecomment-490752199)

@martinisraelsen Hi, sure I will move it there. :-)

**Comment by web-padawan** [comment]
Safari 12 seems to have a bug where the result of `getComputedStyle` for an element doesn't always accurately reflect values from recently changed CSS variables. Reading the element's `offsetWidth` causes a style recalc which seems to update them.

https://bugs.webkit.org/show_bug.cgi?id=170708

Source:
https://github.com/Polymer/polymer/blob/164a7cad43c1f4f75b406557f894ac3a4394972a/test/unit/styling-cross-scope-var.html#L940

**Comment by web-padawan** [comment]
See also https://github.com/vaadin/vaadin-form-layout/issues/111#issue-447558602

**Comment by Haprog** [comment]
The Safari bug mentioned above (related to `getComputedStyle` and `offsetWidth`) doesn't seem to be related to this issue. 

This issue seems to be a semi-randomly occurring rendering bug in Safari which seems to be triggered by this animation style in our CSS:
```css
:host {
  animation: 1ms vaadin-form-layout-appear;
}
@keyframes vaadin-form-layout-appear {
  to {
    opacity: 1;
  }
}
```

This animation serves no visual purpose, but it is used to detect that the element has been rendered (to trigger additional logic) in an `animationend` event listener. It's basically just a dummy animation that triggers an `animationend` event when the element is rendered by the browser.

It seems that in Safari this animating of opacity from `1` to `1` causes weird flickering when the components are being added to DOM and sometimes the components get left rendered in wrong place (even though when inspecting in dev tools they seem to be positioned correctly).

### Reproducing

Here's a pretty minimal HTML page that reproduces the issue easily:

```html
<!doctype html>
<html>
  <head>
    <script src="/bower_components/webcomponentsjs/webcomponents-loader.js"></script>
    <link rel="import" href="/bower_components/vaadin-form-layout/vaadin-form-layout.html">
    <link rel="import" href="/bower_components/vaadin-text-field/vaadin-text-field.html">
  </head>
  <body>
    <div id="target"></div>
    <button id="btnAdd">Add</button>
    <script>
      customElements.whenDefined('vaadin-form-layout').then(function() {
        const html = Polymer.html;
        const target = document.getElementById('target');
        const button = document.getElementById('btnAdd');
        button.addEventListener('click', function() {
          target.appendChild(html`
            <vaadin-form-layout>
              <vaadin-text-field></vaadin-text-field>
            </vaadin-form-layout>`.content
          );
        });
      });
    </script>
  </body>
</html>
```

### Workaround

As mentioned in #111 you can workaround this issue by adding a style of `opacity: 0.99 !important;` to the `vaadin-form-layout`.

Instead of that you can also use e.g. `transform: translateX(0);` so you don't need to use `!important` or actually change opacity.

This workaround is best applied to all form layouts by creating a style module, but it can also be added to specific form layouts by adding it just as an inline style (either in an HTML template or in Java via `formLayout.getStyle().set("transform", "translateX(0)");`)

### Fix

It seems that we can fix this in the component by changing the animation CSS. From quick tests it seems that we can just use another CSS property in the animation (e.g. `visibility: visible;`) instead of `opacity: 1;` to get this working in Safari without problems and still get the `animationend` event we need.

I'll do some more testing to check if the `visibility: visible;` approach for the animation event works consistently in all browsers. If it works fine, I'll commit a fix.

**Comment by Haprog** [comment]
Using `visibility: visible;` seems to work well otherwise, but it seems that when the developer has set `visibility: hidden;` on the component, it can flash visible when adding it to the DOM. This doesn't seem to happen with `opacity: 1;` in the animation even if the developer has set `opacity: 0;` for the component.

I found that a slightly better way to fix it seems to be just adding `!important` to the animation rule so it's:
```css
@keyframes vaadin-form-layout-appear {
  to {
    opacity: 1 !important;
  }
}
```

This actually makes it so that the rule is ignored (by the spec) but it still seems to trigger the `animationend` even we need (in all browsers).

> Declarations in a keyframe qualified with `!important` are ignored.
– https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes#!important_in_a_keyframe

> "None of the properties interact with the cascade (so using !important on them is invalid and will cause the property to be ignored)."
> "properties qualified with !important are invalid and ignored"
– https://drafts.csswg.org/css-animations/#keyframes

Though according to MDN `!important` rules in keyframes are not ignored in IE11 or Edge, but that doesn't cause an issue for us here since `opacity: 1;` was good enough for all other browsers atm. except new Safari which has some rendering bug that this seems to solve.

I tested that using this `opacity: 1 !important;` in the animation triggers the event (even if `opacity: 0;` is set by some other style) and doesn't cause visual issues in: Chrome 74, Firefox 67, Safari 12.1.1, Edge Canary 76, IE11, Edge 17, iOS 9.3 Safari, iOS 12.2 Safari.

**Comment by Haprog** [comment]
The fix for this has been released in `vaadin-form-layout` [v2.1.3](https://github.com/vaadin/vaadin-form-layout/releases/tag/v2.1.3).

**Comment by Haprog** [comment]
The fixed version should be included to next maintenance release of Vaadin 13 and next prerelease of Vaadin 14 but in the meanwhile (if you need this fix right now in a Flow app) you can pin this dependency to your `pom.xml` to use the latest version with the fix:

```xml
<dependency>
    <groupId>org.webjars.bowergithub.vaadin</groupId>
    <artifactId>vaadin-form-layout</artifactId>
    <version>2.1.3</version>
</dependency>
```
